File: crates/speech-engine/tests/protocol.rs
//! The wire between the app and the sidecar.
//!
//! One JSON object per line, over a child process's stdin and stdout. It is the
//! seam most likely to break silently — a renamed field or a changed reply
//! shape shows up as a mysterious failure at generation time, hours after the
//! change that caused it. So these tests drive a stand-in sidecar written for
//! the purpose: they check the protocol, not the model.
//!
//! `sidecar.rs` fails a spawn fast by calling `ping` during construction, which
//! means every case below has to answer that first.

use std::path::{Path, PathBuf};

use speech_engine::sidecar::MlxSidecar;
use speech_engine::{EngineError, SpeechEngine, SynthesisRequest};

/// Write a Python script that speaks the protocol however the test needs, and
/// hand back its path. `body` is the per-request behaviour, with `req` in scope.
fn stand_in(dir: &Path, name: &str, body: &str) -> PathBuf {
    let script = dir.join(format!("{name}.py"));
    let source = format!(
        r#"
import json, sys

def reply(obj):
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()

for line in sys.stdin:
    line = line.strip()
    if not line:
        continue
    req = json.loads(line)
    if req["method"] == "ping":
        reply({{"id": req["id"], "ok": True, "result": {{"pong": True}}}})
        continue
{body}
"#
    );
    std::fs::write(&script, source).expect("write stand-in sidecar");
    script
}

fn python() -> PathBuf {
    PathBuf::from("/usr/bin/python3")
}

fn spawn(dir: &tempfile::TempDir, name: &str, body: &str) -> speech_engine::Result<MlxSidecar> {
    let script = stand_in(dir.path(), name, body);
    MlxSidecar::spawn(&python(), &script, dir.path())
}

fn request() -> SynthesisRequest {
    SynthesisRequest {
        text: "Good morning.".into(),
        output: PathBuf::from("/tmp/out.wav"),
        model: Some("dots-tts-mf".into()),
        voice_id: None,
        seed: Some(4417),
        name: None,
        clip_id: None,
    }
}

#[test]
fn spawn_completes_the_ping_handshake() {
    let dir = tempfile::tempdir().unwrap();
    // A sidecar that answers nothing but ping still has to spawn: the handshake
    // is the whole contract at construction time.
    assert!(spawn(&dir, "ping_only", "    pass").is_ok());
}

#[test]
fn spawn_fails_when_the_interpreter_is_missing() {
    let dir = tempfile::tempdir().unwrap();
    let script = stand_in(dir.path(), "unused", "    pass");
    match MlxSidecar::spawn(Path::new("/nonexistent/python"), &script, dir.path()) {
        Err(EngineError::Transport(_)) => {}
        Err(other) => panic!("expected Transport, got {other:?}"),
        Ok(_) => panic!("a missing interpreter cannot be spawned"),
    }
}

#[test]
fn a_result_is_parsed_into_its_type() {
    let dir = tempfile::tempdir().unwrap();
    let mut engine = spawn(
        &dir,
        "synth",
        r#"    reply({"id": req["id"], "ok": True, "result": {
        "output": req["params"]["output"], "model": "dots-tts-mf",
        "audio_s": 5.44, "gen_s": 12.0, "rtf": 2.2, "seed": 4417,
        "sample_rate": 24000, "clip": None}})"#,
    )
    .unwrap();

    let out = engine.synthesize(&request()).expect("a well-formed reply parses");
    assert_eq!(out.model, "dots-tts-mf");
    assert_eq!(out.seed, Some(4417));
    assert_eq!(out.sample_rate, 24000);
    assert!((out.audio_s - 5.44).abs() < 1e-6);
}

#[test]
fn the_request_carries_every_field_the_engine_needs() {
    let dir = tempfile::tempdir().unwrap();
    // Echo the params back through a field the reply already has, so a dropped
    // or renamed request field fails here rather than at generation time.
    let mut engine = spawn(
        &dir,
        "echo",
        r#"    p = req["params"]
    reply({"id": req["id"], "ok": True, "result": {
        "output": "/tmp/out.wav", "model": json.dumps(sorted(p.keys())),
        "audio_s": 1.0, "gen_s": 1.0, "rtf": 1.0, "seed": p.get("seed"),
        "sample_rate": 24000, "clip": None}})"#,
    )
    .unwrap();

    let out = engine.synthesize(&request()).unwrap();
    let sent: Vec<String> = serde_json::from_str(&out.model).unwrap();
    for field in ["text", "output", "model", "voice_id", "seed", "name", "clip_id"] {
        assert!(sent.contains(&field.to_string()), "{field} was not sent: {sent:?}");
    }
}

#[test]
fn a_clip_of_several_takes_is_parsed() {
    let dir = tempfile::tempdir().unwrap();
    // Generating again adds a take under the same clip, so the reply carries
    // the whole list — newest first — rather than one file.
    let mut engine = spawn(
        &dir,
        "takes",
        r#"    reply({"id": req["id"], "ok": True, "result": {
        "output": "/tmp/out.wav", "model": "dots-tts-mf",
        "audio_s": 1.0, "gen_s": 1.0, "rtf": 1.0, "seed": 2796,
        "sample_rate": 24000,
        "clip": {"id": "clip-1", "title": "t", "name": "n", "text": "t",
                 "voice_id": None, "model": "dots-tts-mf", "created": "now",
                 "takes": [
                   {"id": "take-2", "path": "/tmp/b.wav", "audio_s": 1.3,
                    "gen_s": 4.0, "seed": 2796, "created": "17:02"},
                   {"id": "take-1", "path": "/tmp/a.wav", "audio_s": 1.4,
                    "gen_s": 5.0, "seed": 9895, "created": "13:06"}]}}})"#,
    )
    .unwrap();

    let clip = engine.synthesize(&request()).unwrap().clip.expect("a clip came back");
    assert_eq!(clip.takes.len(), 2);
    assert_eq!(clip.latest().map(|t| t.id.as_str()), Some("take-2"), "newest first");
    assert_eq!(clip.take("take-1").map(|t| t.seed), Some(Some(9895)));
    assert!(clip.take("missing").is_none());
    let rtf = clip.latest().unwrap().rtf().unwrap();
    assert!((rtf - 4.0 / 1.3).abs() < 1e-5, "{rtf}");
}

#[test]
fn a_clip_written_before_takes_existed_still_parses() {
    let dir = tempfile::tempdir().unwrap();
    // The sidecar migrates on read, but the Rust side must not fall over if it
    // ever sees the old shape — `takes` defaults rather than failing.
    let mut engine = spawn(
        &dir,
        "old_clip",
        r#"    reply({"id": req["id"], "ok": True, "result": {
        "output": "/tmp/out.wav", "model": "dots-tts-mf",
        "audio_s": 1.0, "gen_s": 1.0, "rtf": 1.0, "seed": 1,
        "sample_rate": 24000,
        "clip": {"id": "clip-1", "title": "t", "name": "n", "text": "t",
                 "voice_id": None, "model": "dots-tts-mf", "created": "now"}}})"#,
    )
    .unwrap();

    let clip = engine.synthesize(&request()).unwrap().clip.unwrap();
    assert!(clip.takes.is_empty());
    assert!(clip.latest().is_none());
}

#[test]
fn an_engine_side_refusal_is_rejected_not_transport() {
    let dir = tempfile::tempdir().unwrap();
    let mut engine = spawn(
        &dir,
        "refuse",
        r#"    reply({"id": req["id"], "ok": False, "error": "nothing to say"})"#,
    )
    .unwrap();

    // The distinction matters: `Rejected` is the engine answering, and the app
    // shows it. `NotRunning` restarts the process.
    match engine.synthesize(&request()) {
        Err(EngineError::Rejected(message)) => assert_eq!(message, "nothing to say"),
        other => panic!("expected Rejected, got {other:?}"),
    }
}

#[test]
fn a_reply_of_the_wrong_shape_is_a_transport_error() {
    let dir = tempfile::tempdir().unwrap();
    let mut engine = spawn(
        &dir,
        "wrong_shape",
        r#"    reply({"id": req["id"], "ok": True, "result": {"audio_s": "five"}})"#,
    )
    .unwrap();

    match engine.synthesize(&request()) {
        Err(EngineError::Transport(message)) => {
            assert!(message.contains("unexpected result shape"), "{message}")
        }
        other => panic!("expected Transport, got {other:?}"),
    }
}

#[test]
fn a_line_that_is_not_json_is_a_transport_error() {
    let dir = tempfile::tempdir().unwrap();
    let mut engine = spawn(
        &dir,
        "garbage",
        r#"    sys.stdout.write("this is not json\n"); sys.stdout.flush()"#,
    )
    .unwrap();

    match engine.synthesize(&request()) {
        Err(EngineError::Transport(message)) => {
            assert!(message.contains("unparseable reply"), "{message}")
        }
        other => panic!("expected Transport, got {other:?}"),
    }
}

#[test]
fn a_sidecar_that_exits_reports_not_running() {
    let dir = tempfile::tempdir().unwrap();
    // Answers ping, then leaves. The app restarts the engine on this and only
    // on this, so it must not be reported as a transport failure.
    let mut engine = spawn(&dir, "exits", r#"    sys.exit(0)"#).unwrap();

    match engine.synthesize(&request()) {
        Err(EngineError::NotRunning) => {}
        other => panic!("expected NotRunning, got {other:?}"),
    }
}

#[test]
fn ids_increase_so_replies_cannot_be_read_out_of_order() {
    let dir = tempfile::tempdir().unwrap();
    let mut engine = spawn(
        &dir,
        "ids",
        r#"    reply({"id": req["id"], "ok": True, "result": {
        "output": "/tmp/out.wav", "model": str(req["id"]),
        "audio_s": 1.0, "gen_s": 1.0, "rtf": 1.0, "seed": 1,
        "sample_rate": 24000, "clip": None}})"#,
    )
    .unwrap();

    let first: u64 = engine.synthesize(&request()).unwrap().model.parse().unwrap();
    let second: u64 = engine.synthesize(&request()).unwrap().model.parse().unwrap();
    assert!(second > first, "ids went {first} then {second}");
}

#[test]
fn dropping_the_engine_stops_the_child() {
    let dir = tempfile::tempdir().unwrap();
    let marker = dir.path().join("still-alive");
    // The child writes a marker as it goes down cleanly; killed, it never gets
    // to. Either way the point is that the process does not outlive the engine.
    let engine = spawn(
        &dir,
        "long_lived",
        &format!(
            "    import time\n    time.sleep(30)\n    open({:?}, 'w').write('x')",
            marker.display().to_string()
        ),
    )
    .unwrap();
    drop(engine);

    // A killed child releases its end of the pipe immediately; give the OS a
    // moment, then check nothing wrote the marker.
    std::thread::sleep(std::time::Duration::from_millis(200));
    assert!(!marker.exists(), "the sidecar outlived the engine that owned it");
}

#[test]
fn stdout_noise_before_the_reply_is_not_tolerated_silently() {
    let dir = tempfile::tempdir().unwrap();
    // A print() left in the sidecar lands on stdout ahead of the reply. That is
    // a real mistake to make, and it must fail loudly rather than hang or
    // return the wrong thing.
    let mut engine = spawn(
        &dir,
        "chatty",
        r#"    print("loading model")
    reply({"id": req["id"], "ok": True, "result": {
        "output": "/tmp/out.wav", "model": "dots-tts-mf",
        "audio_s": 1.0, "gen_s": 1.0, "rtf": 1.0, "seed": 1,
        "sample_rate": 24000, "clip": None}})"#,
    )
    .unwrap();

    match engine.synthesize(&request()) {
        Err(EngineError::Transport(_)) => {}
        other => panic!("noise on stdout is a protocol violation, got {other:?}"),
    }
}

/// The real sidecar, against the real protocol. Ignored by default because it
/// needs the installed runtime; run it with `cargo test -- --ignored` on a
/// machine that has one.
#[test]
#[ignore = "needs the installed Python runtime"]
fn the_real_sidecar_answers_a_ping() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR")).join("../..");
    let paths = speech_engine::EnginePaths::resolve(&root);
    if let Some(missing) = paths.missing() {
        panic!("{missing}");
    }

    let mut engine = MlxSidecar::spawn(&paths.python, &paths.script, &paths.work_dir)
        .expect("spawn the real sidecar");
    let models = engine.models(false).expect("the real sidecar lists models");
    assert!(!models.is_empty(), "the catalogue is empty");
}

File: sidecar/protocol.py
"""Serving the Yarngo engine API: JSON-RPC 2.0, one object per line, over stdio.

The wire is not ours: request ids that correlate replies independently of
order, notifications that expect no reply, and a defined error shape are all
JSON-RPC 2.0, and reinventing them would only mean getting them slightly wrong.
The methods are ours, and so is everything below — which is the part JSON-RPC
says nothing about.

The previous version read one request, ran it to completion, and wrote one
reply. While it worked it was not reading, so nothing could be asked and
nothing could be said — which is why progress and cancellation ended up
travelling by file, and why a cheap call waited behind a long one.

Three parts, each owning one thing:

    reader      the main thread. Parses, and decides where a request goes.
                It never touches the model, because it must never be slow.
    actor       one worker thread. Owns the model and everything derived from
                it, so nothing else can read a cache while it is being written.
    writer      one thread. The only thing that writes to stdout, so two
                answers cannot interleave halves of a line.

Requests are sorted by what they own, not by how long they take. A method that
looks cheap but reads model state still belongs to the actor: `list_voices`
reading the voice table while `register_voice` writes it is a torn read that no
timing test would catch.

What can overtake what: a broker reply can arrive before a model reply that was
asked for first, which is the whole point. Model operations cannot overtake each
other — the actor is one thread and runs them in the order they were queued.
"""

from __future__ import annotations

import json
import os
import queue
import sys
import threading
import time
import traceback
from collections import deque
from dataclasses import dataclass, field
from typing import Any, Callable, Protocol

JSONRPC_VERSION = "2.0"

# What this speaks, named so a peer can tell it apart from anything else that
# also happens to be JSON-RPC over a pipe. The wire is a standard; the methods
# are ours, and this is the version of those. It starts at one because nothing
# has shipped: there is no earlier version to be compatible with.
PROTOCOL_NAME = "yarngo-engine"
PROTOCOL_VERSION = 1

# Requests waiting for the model actor: one running, one ready to start. Shallow
# on purpose. A deep queue here would make this a second scheduler holding a
# backlog the application cannot see, reorder, or recover — and the application
# is the one that knows what a restart should retry. Deeper than one so the
# engine is not idle between operations; no deeper, because nothing here can be
# rescheduled.
ACTOR_QUEUE_DEPTH = 2

# How much may be waiting on the control lane. Bounded, and its ceiling is
# calculable rather than hoped for: a reply exists only because a request was
# read, admission is bounded by the actor's queue, and terminal events are one
# per execution. Reaching this means an assumption is wrong, so it is fatal
# rather than dropped — a lost reply strands a caller for ever.
CONTROL_LANE_LIMIT = 512

# How many finished executions to remember, so a cancellation arriving after the
# work ended can say so rather than calling it unknown. Bounded and not
# persisted: this is a courtesy for late messages, never a record. The
# application keeps the history, and after a restart an old execution is
# genuinely unknown here — which is the truth, since whatever was running died
# with the process.
FINISHED_MEMORY = 256

# The longest line either side will accept, enforced while reading rather than
# after. A peer sending more than this is not one of ours, and finding the end
# of its line before objecting would mean allocating whatever it claimed.
MAX_FRAME_BYTES = 8 * 1024 * 1024

# JSON-RPC's own codes, for faults in the exchange itself.
PARSE_ERROR = -32700
INVALID_REQUEST = -32600
METHOD_NOT_FOUND = -32601
INVALID_PARAMS = -32602
INTERNAL_ERROR = -32603

# Ours, in the range JSON-RPC leaves to the server. Numbers rather than message
# matching: the caller decides what to do from these, and rewording a message
# must not change behaviour.
ENGINE_BUSY = -32001
VOICE_DELETED = -32002
MODEL_NOT_INSTALLED = -32003
JOB_ALREADY_TERMINAL = -32004


# What a model operation is: parameters, and the context it reports against.
Handler = Callable[[dict, "Context"], "dict | None"]
# What a broker operation is: parameters alone. Nothing it answers runs long
# enough to have progress to report or a cancellation to honour.
BrokerHandler = Callable[[dict], "dict | None"]


class Cancelled(Exception):
    """Raised by a handler that stopped because it was asked to.

    Not a failure: the work did not finish, and that is the outcome the caller
    asked for. Reported as a cancellation with its own terminal event, because a
    job sitting in `cancel_requested` is waiting to be told how it ended and an
    error would tell it the wrong thing.
    """


class Emit(Protocol):
    def __call__(self, method: str, params: dict) -> None: ...


@dataclass
class Cancellation:
    """Which executions have been asked to stop.

    Keyed on the execution rather than the job. A job may be attempted more than
    once — after an engine restart, or a retry — and a cancellation aimed at the
    attempt that was abandoned must not reach the one that replaced it. Rust
    owns which attempt is current; this only has to answer about the one it was
    told to run.

    Written by the reader thread the moment a cancellation arrives, read by the
    actor at whatever checkpoint it reaches next. That gap is the honest cost of
    cooperative cancellation: the request is acknowledged immediately, and the
    work stops when it can.
    """

    _asked: set[str] = field(default_factory=set)
    _lock: threading.Lock = field(default_factory=threading.Lock)

    def request(self, execution_id: str) -> None:
        with self._lock:
            self._asked.add(execution_id)

    def is_requested(self, execution_id: str) -> bool:
        with self._lock:
            return execution_id in self._asked

    def forget(self, execution_id: str) -> None:
        with self._lock:
            self._asked.discard(execution_id)


@dataclass
class Context:
    """What a model operation is given beyond its parameters.

    Both identifiers travel with everything it says. The job is what the user
    started and what Rust keeps; the execution is this attempt at it. An event
    naming only the job could be mistaken for the current attempt when it came
    from an abandoned one.
    """

    emit_raw: Emit
    cancellation: Cancellation
    job_id: str | None = None
    execution_id: str | None = None

    def emit(self, method: str, params: dict) -> None:
        stamped = dict(params)
        if self.job_id is not None:
            stamped.setdefault("job_id", self.job_id)
        if self.execution_id is not None:
            stamped.setdefault("execution_id", self.execution_id)
        self.emit_raw(method, stamped)

    def cancelled(self) -> bool:
        return self.execution_id is not None and self.cancellation.is_requested(
            self.execution_id
        )


class Writer:
    """The only thing that writes to stdout, in two lanes.

    A bounded queue, never blocking, and never losing anything cannot all hold
    at once, so this says which gives. Progress is coalesced: one entry per
    execution, replaced as newer values arrive, because it is the same fact at
    successive values and only the newest is worth sending. Its memory is
    therefore the number of live executions, not the number of updates.

    The control lane — replies, errors, terminal events — loses nothing. It is
    bounded, but its ceiling is calculable rather than hoped for: a reply exists
    only because a request was read, admission is bounded by the actor's queue,
    and a terminal event is one per execution. Reaching the limit means one of
    those is untrue, and a dropped reply would strand a caller silently, so it
    ends the session instead and lets the parent start a fresh one.

    Nothing that hands it a message waits. Blocking would stall the actor
    mid-operation or the reader mid-request, and a reader that has stopped
    reading cannot receive whatever would have released it.
    """

    def __init__(self, stream, on_fatal=None) -> None:
        self._stream = stream
        self._control: deque = deque()
        self._progress: dict[str, dict] = {}
        self._ready = threading.Condition()
        self._on_fatal = on_fatal or _die
        self._closing = False
        self.coalesced = 0
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def send(self, message: dict) -> None:
        key = _progress_key(message)
        with self._ready:
            if key is not None:
                if key in self._progress:
                    self.coalesced += 1
                self._progress[key] = message
            else:
                if len(self._control) >= CONTROL_LANE_LIMIT:
                    self._on_fatal(
                        f"control lane exceeded {CONTROL_LANE_LIMIT} messages; "
                        "a reply would have to be dropped"
                    )
                    return
                self._control.append(message)
            self._ready.notify()

    def close(self, timeout: float = 5.0) -> None:
        """Wait for what is queued to be written.

        The writer is a daemon thread, so without this the process can end
        between a reply being handed over and it reaching the pipe — and the
        caller waiting on that reply learns nothing except that the engine
        stopped.
        """
        with self._ready:
            self._closing = True
            self._ready.notify()
        self._thread.join(timeout)

    def _run(self) -> None:
        while True:
            with self._ready:
                while not self._control and not self._progress and not self._closing:
                    self._ready.wait()
                if self._closing and not self._control and not self._progress:
                    return
                # Control first and in full: progress that is superseded while
                # a reply is being written was never worth sending.
                batch = list(self._control)
                self._control.clear()
                batch.extend(self._progress.values())
                self._progress.clear()
            for message in batch:
                self._stream.write(json.dumps(message) + "\n")
                self._stream.flush()

    def reply(self, request_id: Any, result: dict) -> None:
        self.send({"jsonrpc": JSONRPC_VERSION, "id": request_id, "result": result})

    def error(self, request_id: Any, code: int, message: str, data: dict | None = None) -> None:
        payload = {"code": code, "message": message}
        if data is not None:
            payload["data"] = data
        self.send({"jsonrpc": JSONRPC_VERSION, "id": request_id, "error": payload})

    def event(self, method: str, params: dict) -> None:
        """A notification: no id, and no reply is expected or permitted."""
        self.send({"jsonrpc": JSONRPC_VERSION, "method": method, "params": params})


class ModelActor:
    """One thread, and the only one that may touch the model.

    Everything derived from a model — the loaded weights, the conditioning
    cache, the generator — is reachable from exactly here. Operations queue
    rather than run concurrently, which is a deliberate limit: a desktop
    accelerator gains nothing from two generations at once, and the caches
    behind them are not written to be shared.
    """

    def __init__(self, handlers: dict[str, Handler], writer: Writer, cancellation: Cancellation) -> None:
        self._handlers = handlers
        self._writer = writer
        self._cancellation = cancellation
        self._work: queue.Queue = queue.Queue(maxsize=ACTOR_QUEUE_DEPTH)
        # What this engine was told to run, so a cancellation can be answered
        # with what happened rather than with an acknowledgement either way.
        # Knowing its own executions is not scheduling: the caller decides what
        # to run and in what order.
        self._lock = threading.Lock()
        self._queued: set[str] = set()
        self._running: str | None = None
        # Whether a handler is in progress, which is not the same question as
        # which execution is in progress: most operations have no execution at
        # all, and `_running is None` is their normal state while working.
        self._busy = False
        self._finished: deque = deque(maxlen=FINISHED_MEMORY)
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def state_of(self, execution_id: str) -> str:
        with self._lock:
            if execution_id == self._running:
                return "running"
            if execution_id in self._queued:
                return "queued"
            if execution_id in self._finished:
                return "terminal"
        return "unknown"

    def submit(
        self, request_id: Any, method: str, params: dict, is_notification: bool = False
    ) -> bool:
        """Queue an operation. False when the queue is full, so the caller can
        be refused rather than left waiting on a reply that will not come."""
        execution_id = params.get("execution_id") if isinstance(params, dict) else None
        try:
            self._work.put_nowait((request_id, method, params, is_notification))
        except queue.Full:
            return False
        if execution_id:
            with self._lock:
                self._queued.add(execution_id)
        return True

    def depth(self) -> int:
        return self._work.qsize()

    def drain(self, timeout: float = 30.0) -> None:
        """Wait for what has been accepted to finish.

        Input ending does not end the work already taken from it. Closing the
        writer first would lose the answer to a request that was accepted, which
        is the one kind of silence a caller cannot recover from.
        """
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            with self._lock:
                idle = not self._busy
            if idle and self._work.empty():
                return
            time.sleep(0.01)

    def _run(self) -> None:
        while True:
            request_id, method, params, is_notification = self._work.get()
            fields = params if isinstance(params, dict) else {}
            job_id = fields.get("job_id")
            execution_id = fields.get("execution_id")
            with self._lock:
                self._queued.discard(execution_id)
                self._running = execution_id
                self._busy = True
            # Checked before the handler is entered, not only inside it: an
            # execution cancelled while it waited its turn should never start.
            if execution_id and self._cancellation.is_requested(execution_id):
                self._cancellation.forget(execution_id)
                with self._lock:
                    self._running = None
                    self._busy = False
                    self._finished.append(execution_id)
                # The terminal event, not only the reply. The caller's job is
                # sitting in `cancel_requested` waiting to be told how it ended,
                # and a reply to the submission does not say that: the two are
                # answers to different questions.
                self._writer.event(
                    "job.cancelled",
                    {"job_id": job_id, "execution_id": execution_id, "started": False},
                )
                if not is_notification:
                    self._writer.reply(request_id, {"state": "cancelled", "started": False})
                continue
            context = Context(
                emit_raw=self._writer.event,
                cancellation=self._cancellation,
                job_id=job_id,
                execution_id=execution_id,
            )
            try:
                result = self._handlers[method](params, context)
                if not is_notification:
                    self._writer.reply(request_id, result if result is not None else {})
            except Cancelled as stopped:
                # Started and then stopped, which is a different fact from
                # stopped before starting: work was done and any partial output
                # exists. Both say `started` so the caller can tell them apart.
                self._writer.event(
                    "job.cancelled",
                    {
                        "job_id": job_id,
                        "execution_id": execution_id,
                        "started": True,
                        "detail": str(stopped),
                    },
                )
                if not is_notification:
                    self._writer.reply(request_id, {"state": "cancelled", "started": True})
            except Exception as exc:  # noqa: BLE001
                # Reported as a reply, never as a stack trace on stdout: the
                # caller is waiting on this id and gets an error rather than a
                # timeout, and the trace goes to the log.
                _log(traceback.format_exc())
                if not is_notification:
                    self._writer.error(
                        request_id, INTERNAL_ERROR, f"{type(exc).__name__}: {exc}"
                    )
            finally:
                with self._lock:
                    self._running = None
                    self._busy = False
                    if execution_id:
                        self._finished.append(execution_id)
                if execution_id:
                    self._cancellation.forget(execution_id)


def _log(message: str) -> None:
    print(message, file=sys.stderr, flush=True)


def _die(reason: str) -> None:
    """End the session. The parent notices the output stop and starts another."""
    _log(f"protocol failure: {reason}")
    sys.stderr.flush()
    os._exit(70)


def _progress_key(message: dict) -> str | None:
    """Which stream of progress this supersedes, or None if it supersedes
    nothing. Keyed per execution so one job's updates cannot displace another's."""
    if message.get("method") != "job.progress":
        return None
    params = message.get("params") or {}
    return str(params.get("execution_id") or params.get("job_id") or "")


def serve(
    *,
    broker: dict[str, BrokerHandler],
    model: dict[str, Handler],
    capabilities: dict | None = None,
    stdin=None,
    stdout=None,
    on_fatal=None,
) -> None:
    """Read requests until the input ends.

    `broker` methods run on this thread and must not touch model state.
    `model` methods run on the actor, one at a time.
    """
    # The protocol keeps the real stdout, and everything else is pointed at the
    # log. A library writing a progress bar to stdout would otherwise put a line
    # on the wire that is not a frame, and one such line is enough.
    protocol_out = stdout if stdout is not None else sys.stdout
    if stdout is None:
        sys.stdout = sys.stderr

    source = stdin if stdin is not None else sys.stdin
    on_fatal = on_fatal or _die
    writer = Writer(protocol_out, on_fatal=on_fatal)
    cancellation = Cancellation()
    actor = ModelActor(model, writer, cancellation)

    while True:
        # Read at most one frame's worth. A line longer than the limit comes
        # back without its newline, which is how an oversized frame is told
        # apart from a large-but-legal one.
        raw = source.readline(MAX_FRAME_BYTES + 1)
        if raw == "":
            break
        if len(raw) > MAX_FRAME_BYTES and not raw.endswith("\n"):
            # No attempt to resynchronise: the rest of that frame is still in
            # the stream, and guessing where it ends is how a parser starts
            # reading someone else's bytes as a message.
            writer.error(None, INVALID_REQUEST, "frame exceeds the size limit")
            on_fatal("oversized frame")
            return
        line = raw.strip()
        if not line:
            continue
        try:
            request = json.loads(line)
        except json.JSONDecodeError as exc:
            # A parse failure has no id to answer against, and null is what the
            # specification asks for in exactly this case.
            writer.error(None, PARSE_ERROR, f"bad json: {exc}")
            continue

        problem = _envelope_fault(request)
        if problem is not None:
            writer.error(
                request.get("id") if isinstance(request, dict) else None,
                INVALID_REQUEST,
                problem,
            )
            continue

        method = request["method"]
        # Kept as sent. `or {}` turned an empty positional list into an object,
        # which is a different call.
        params = request["params"] if "params" in request else {}
        if params is None:
            params = {}
        # Absent rather than null: a request without an id is a notification,
        # and answering one is as wrong as failing to answer a request.
        request_id = request.get("id")
        is_notification = "id" not in request

        if method == "initialize":
            if not is_notification:
                writer.reply(
                    request_id,
                    {
                        "protocol": PROTOCOL_NAME,
                        "version": PROTOCOL_VERSION,
                        **(capabilities or {}),
                    },
                )
            continue

        if method == "job.cancel":
            # Answered here rather than queued, which is the point of it: a
            # cancellation that waited its turn behind the job it is cancelling
            # would arrive after the work it was meant to stop.
            # By execution rather than by job: the caller knows which attempt is
            # current, and a cancellation meant for an abandoned one must not
            # reach its replacement.
            execution_id = params.get("execution_id") if isinstance(params, dict) else None
            if not execution_id:
                if not is_notification:
                    writer.error(
                        request_id, INVALID_PARAMS, "job.cancel needs an execution_id"
                    )
            else:
                # Answered with what is true of this execution. A blanket
                # acknowledgement would tell a caller its cancellation landed
                # when the work had already finished, or never existed.
                state = actor.state_of(execution_id)
                if state in ("queued", "running"):
                    cancellation.request(execution_id)
                outcome = {
                    "queued": "cancel_requested",
                    "running": "cancel_requested",
                    "terminal": "already_terminal",
                    "unknown": "unknown_execution",
                }[state]
                if not is_notification:
                    writer.reply(request_id, {"state": outcome, "execution_state": state})
            continue

        if method in broker:
            try:
                result = broker[method](params) or {}
                if not is_notification:
                    writer.reply(request_id, result)
            except Exception as exc:  # noqa: BLE001
                _log(traceback.format_exc())
                if not is_notification:
                    writer.error(request_id, INTERNAL_ERROR, f"{type(exc).__name__}: {exc}")
            continue

        if method in model:
            if not actor.submit(request_id, method, params, is_notification):
                if not is_notification:
                    writer.error(
                        request_id,
                        ENGINE_BUSY,
                        "the engine has too much queued",
                        {"queue_depth": actor.depth()},
                    )
            continue

        if not is_notification:
            writer.error(request_id, METHOD_NOT_FOUND, f"unknown method {method!r}")

    # The input has ended, so nothing more will be asked. What was already
    # accepted still has to finish, and what it answers still has to arrive.
    actor.drain()
    writer.close()


def _envelope_fault(request: Any) -> str | None:
    """Why this is not a JSON-RPC request, or None when it is one."""
    if isinstance(request, list):
        # A valid JSON-RPC construction this profile does not serve. Refused by
        # name, so a caller learns which it is rather than guessing from a
        # generic complaint about shape.
        return "batches are not supported by this engine"
    if not isinstance(request, dict):
        return "a request must be an object"
    if request.get("jsonrpc") != JSONRPC_VERSION:
        return f"expected jsonrpc {JSONRPC_VERSION!r}, got {request.get('jsonrpc')!r}"
    method = request.get("method")
    if not isinstance(method, str):
        return "method must be a string"
    params = request.get("params")
    if params is not None and not isinstance(params, (dict, list)):
        return "params must be an object or an array"
    if "id" in request and not isinstance(request["id"], (str, int, float, type(None))):
        return "id must be a string, a number, or null"
    return None
